e2e/servicediscovery/service_discovery_test.go

```diff
@@ -0,0 +1,353 @@
+// such as Consul Connect, which have their own test suite.
+func TestServiceDiscovery(t *testing.T) {
+
+	// Wait until we have a usable cluster before running the tests.
+	nomadClient := e2eutil.NomadClient(t)
+	e2eutil.WaitForLeader(t, nomadClient)
+	e2eutil.WaitForNodesReady(t, nomadClient, 1)
+
+	// Run our test cases.
+	t.Run("TestServiceDiscovery_MultiProvider", testMultiProvider)
+	t.Run("TestServiceDiscovery_UpdateProvider", testUpdateProvider)
+}
+
+// testMultiProvider tests service discovery where multi providers are used
+// within a single job.
+func testMultiProvider(t *testing.T) {
+
+	nomadClient := e2eutil.NomadClient(t)
+	consulClient := e2eutil.ConsulClient(t)
+
+	// Generate our job ID which will be used for the entire test.
+	jobID := "service-discovery-multi-provider-" + uuid.Generate()[:8]
+	jobIDs := []string{jobID}
+
+	// Defer a cleanup function to remove the job. This will trigger if the
+	// test fails, unless the cancel function is called.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer e2eutil.CleanupJobsAndGCWithContext(t, ctx, &jobIDs)
+
+	// Register the job which contains two groups, each with a single service
+	// that use different providers.
+	allocStubs := e2eutil.RegisterAndWaitForAllocs(t, nomadClient, jobMultiProvider, jobID, "")
+	require.Len(t, allocStubs, 2)
+
+	// We need to understand which allocation belongs to which group so we can
+	// test the service registrations properly.
+	var nomadProviderAllocID, consulProviderAllocID string
+
+	for _, allocStub := range allocStubs {
+		switch allocStub.TaskGroup {
+		case "service_discovery":
+			consulProviderAllocID = allocStub.ID
+		case "service_discovery_secondary":
+			nomadProviderAllocID = allocStub.ID
+		default:
+			t.Fatalf("unknown task group allocation found: %q", allocStub.TaskGroup)
+		}
+	}
+
+	require.NotEmpty(t, nomadProviderAllocID)
+	require.NotEmpty(t, consulProviderAllocID)
+
+	// Services are registered on by the client, so we need to wait for the
+	// alloc to be running before continue safely.
+	e2eutil.WaitForAllocsRunning(t, nomadClient, []string{nomadProviderAllocID, consulProviderAllocID})
+
+	// Lookup the service registration in Nomad and assert this matches what we
+	// expected.
+	require.Eventually(t, func() bool {
+		services, _, err := nomadClient.Services().Get("http-api-nomad", nil)
+		if err != nil {
+			return false
+		}
+
+		// Perform the checks.
+		if len(services) != 1 {
+			return false
+		}
+		if services[0].ServiceName != "http-api-nomad" {
+			return false
+		}
+		if services[0].Namespace != api.DefaultNamespace {
+			return false
+		}
+		if services[0].Datacenter != "dc1" {
+			return false
+		}
+		if services[0].JobID != jobID {
+			return false
+		}
+		if services[0].AllocID != nomadProviderAllocID {
+			return false
+		}
+		return reflect.DeepEqual(services[0].Tags, []string{"foo", "bar"})
+	}, defaultWaitForTime, defaultTickTime)
```

Comment:
This pattern appears often (and I suspect future tests will also use it), would it be possible to pull this out into a helper function?

Reply:
Good shout! I'll get that sorted shortly.